Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift

```diff
@@ -925,28 +925,116 @@ extension SwiftLanguageServer {
+      skreq[keys.compilerargs] = settings.compilerArguments
+    }
+
+    let handle = sourcekitd.send(skreq) { [weak self] result in
+      guard let self = self else {
+        completion([])
+        return
+      }
+      guard let dict = result.success else {
+        log("failed to find refactor actions: \(result.failure!)")
+        completion([])
+        return
+      }
+      guard let results: SKResponseArray = dict[self.keys.refactor_actions] else {
+        completion([])
+        return
+      }
+      var codeActions = [CodeAction]()
+      results.forEach { _, value in
+        if let name: String = value[self.keys.actionname],
+           let actionuid: sourcekitd_uid_t = value[self.keys.actionuid],
+           let ptr = self.sourcekitd.api.uid_get_string_ptr(actionuid)
+        {
+          let actionName = String(cString: ptr)
+          guard actionName != "source.refactoring.kind.rename.global" else {
+            //TODO: Global refactoring
+            return true
+          }
+          let refactorCommand = SemanticRefactorCommand(title: name,
+                                                        actionString: actionName,
+                                                        positionRange: params.range.asRange,
+                                                        textDocument: params.textDocument)
+          do {
+            let lspCommand = try refactorCommand.asCommand()
+            let codeAction = CodeAction(title: name, kind: .refactor, command: lspCommand)
+            codeActions.append(codeAction)
+          } catch {
+            log("Failed to convert SwiftCommand to Command type: \(error)")
+          }
+        }
+        return true
+      }
+
+      completion(codeActions)
+    }
+
+    // FIXME: cancellation
+    _ = handle
+  }
+
   func executeCommand(_ req: Request<ExecuteCommandRequest>) {
-    //TODO: Implement commands.
-    return req.reply(nil)
+    let params = req.params
+    //TODO: If there's support for several types of commands, we might need to structure this similarly to the code actions request.
+    guard let swiftCommand = params.swiftCommand(ofType: SemanticRefactorCommand.self) else {
+      log("semantic refactoring: unknown command \(params.command)", level: .warning)
+      return req.reply(nil)
+    }
+    let url = swiftCommand.textDocument.url
+    semanticRefactoring(swiftCommand) { result in
+      switch result {
+      case .success(let refactor):
+        let edit = refactor.edit
+        self.applyEdit(label: refactor.title, edit: edit) {
+          req.reply(edit)
+        }
+      case .failure(let error):
+        log("semantic refactoring failed \(url): \(error)", level: .warning)
+        return req.reply(nil)
+      }
+    }
   }
 
-  func applyEdit(label: String, edit: WorkspaceEdit) {
+  func applyEdit(label: String, edit: WorkspaceEdit, completion: @escaping () -> Void) {
```

Comment:
Should the completion handler pass through any errors so that failing to apply the edit can fail the executeCommand request?